### DISCUSSION TITLE
How do you enable code coverage?

### DISCUSSION THREAD

**Discussion by StanBarrows** [beyondcode/herd-community#125] [Q&A]
How do you enable code coverage?

In my Herd php.ini `xdebug.mode=coverage` is enabled. But if I run my tests with the --coverage flag I'm receiving an error.

```
 php artisan test --coverage

 ERROR  Code coverage driver not available. Did you install Xdebug or PCOV?
```

**Comment by chongkan** [comment]
The PHP.ini file is located in the Application Support dir. 
https://stackoverflow.com/questions/11829931/setting-xdebug-coverage-enable-on-on-command-line-for-phpunit

**Reply by chongkan** [reply]
PS... I was only able to get that one project setup, since then, I have tried for days and hours.. 1000s of different configs around the web and according to herd, I need to run "herd coverage"... where? with what config? 

about to start building a Docker fot this project and future projects.

**Comment by kauffinger** [comment]
What worked for me (Herd Pro) was moving xdebug config in the main php.ini. With that, coverage works.

**Reply by chongkan** [reply]
moving it? from where to where? 

I ended up setting up a Docker for that project but still coverage wasn't working. Everytime I need to set up coverage, I spend a lot time, days, getting it configured. One finds countless outdated configurations everywhere and wastes a lot of time on that. 

I have a project where I need to stop herd, that makes the SSL config really easy, and run sail in order to have coverage reporting in the local machine. 

For the most recent project, I spent the weekend getting my neck contracted and put in pause yet again to move on to other stuff. Currently leveraging Codecov report in the cloud for that.

**Reply by kauffinger** [reply]
In herd pro, a debug.ini is created automatically which, for me, did nothing to actually enable xdebug coverage mode.
When adding the configuration to the main php.ini for the php versions my projects use, I was able to run pest with coverage enabled.

Not sure what exactly you are trying to accomplish, but that's a way of looking into coverage.

**Comment by alimranahmed** [comment]
I had similar issue and solved it, I used pcov instead of xdebug. It seems lightweight to me. 
I wrote an article on it: [How to install PCOV in Laravel Herd to see code coverage of Laravel application](https://imranic.com/article/how-to-install-pcov-in-laravel-herd-to-see-code-coverage-of-laravel-application)

**Reply by sheadawson** [reply]
Great article! Thank you Al 🙏

**Comment by tomcoonen** [comment]
Ran into this discussion, it's implemented in Herd now: `herd coverage ./vendor/bin/pest --coverage`.
https://herd.laravel.com/docs/1/herd-pro/xdebug#code-coverage

**Comment by JohnnyWalkerDigital** [comment]
None of the above settings worked for me and PHPUnit, but the following did in my php.ini. 

For me it was as simple as adding "coverage" to `xdebug.mode`

> zend_extension=/Applications/Herd.app/Contents/Resources/xdebug/xdebug-84-arm64.so
> xdebug.mode=debug,develop,coverage
> xdebug.start_with_request=yes
> xdebug.start_upon_error=yes
> xdebug.log_level=0
